src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -554,6 +554,62 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
   CallGenerator::do_late_inline_helper();
 }
 
+static bool has_non_debug_usages(Node* n) {
+  for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+    Node* m = n->fast_out(i);
+    if (!m->is_SafePoint()
+        || (m->is_Call() && m->as_Call()->has_non_debug_use(n))) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static void scalarize_debug_usages(CallNode* call, Node* resproj) {
+  Unique_Node_List safepoints;
+  for (DUIterator_Fast imax, i = resproj->fast_outs(imax); i < imax; i++) {
```

Comment:
No need to collect safepoints anymore. You can process users while iterating over them.
```
for (DUIterator_Last imin, i = res->last_outs(imin); i >= imin;) {
  SafePointNode* sfpt = res->last_out(i)->as_SafePoint();
  ...
  int num_edges = sfpt->replace_edges_in_range(res, sobj, start, end);  
  i -= num_edges;
}
```